File: python/vedadb_orm/aggregation.py
"""
VedaDB ORM — AggregationBuilder.

Provides a fluent API for GROUP BY / HAVING / aggregate-function queries.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .exceptions import QueryError
from .utils import format_where_value

if TYPE_CHECKING:
    from .session import Session


class AggregationBuilder:
    """Fluent builder for aggregate queries (COUNT, SUM, AVG, etc.)."""

    def __init__(self, model_cls: type) -> None:
        self._model_cls = model_cls
        self._group_fields: List[str] = []
        self._having_clauses: List[str] = []
        self._aggregates: List[str] = []
        self._where_clauses: List[str] = []
        self._order_clauses: List[str] = []
        self._limit_val: Optional[int] = None
        self._session: Optional["Session"] = None

    # -- grouping -----------------------------------------------------------

    def group_by(self, *fields: str) -> "AggregationBuilder":
        self._group_fields.extend(fields)
        return self

    def having(self, field: str, op: str, value: Any) -> "AggregationBuilder":
        self._having_clauses.append(f"{field} {op} {format_where_value(value)}")
        return self

    # -- aggregate functions ------------------------------------------------

    def count(self, field: str = "*", alias: Optional[str] = None) -> "AggregationBuilder":
        alias = alias or f"count_{field}".replace("*", "all")
        self._aggregates.append(f"COUNT({field}) AS {alias}")
        return self

    def sum(self, field: str, alias: Optional[str] = None) -> "AggregationBuilder":
        alias = alias or f"sum_{field}"
        self._aggregates.append(f"SUM({field}) AS {alias}")
        return self

    def avg(self, field: str, alias: Optional[str] = None) -> "AggregationBuilder":
        alias = alias or f"avg_{field}"
        self._aggregates.append(f"AVG({field}) AS {alias}")
        return self

    def min(self, field: str, alias: Optional[str] = None) -> "AggregationBuilder":
        alias = alias or f"min_{field}"
        self._aggregates.append(f"MIN({field}) AS {alias}")
        return self

    def max(self, field: str, alias: Optional[str] = None) -> "AggregationBuilder":
        alias = alias or f"max_{field}"
        self._aggregates.append(f"MAX({field}) AS {alias}")
        return self

    # -- filtering / ordering / paging --------------------------------------

    def where(self, **conditions: Any) -> "AggregationBuilder":
        for key, val in conditions.items():
            self._where_clauses.append(f"{key} = {format_where_value(val)}")
        return self

    def order_by(self, field: str, direction: str = "ASC") -> "AggregationBuilder":
        direction = direction.upper()
        if direction not in ("ASC", "DESC"):
            direction = "ASC"
        self._order_clauses.append(f"{field} {direction}")
        return self

    def limit(self, n: int) -> "AggregationBuilder":
        self._limit_val = n
        return self

    def use_session(self, session: "Session") -> "AggregationBuilder":
        self._session = session
        return self

    # -- terminal -----------------------------------------------------------

    def exec(self) -> List[Dict[str, Any]]:
        """Execute the aggregation and return a list of dicts."""
        sql = self.to_sql()
        result = self._run_sql(sql)
        return [dict(zip(result.columns, row)) for row in result.rows]

    def to_sql(self) -> str:
        """Generate the SQL string without executing it."""
        table = self._model_cls._table_name
        select_parts: List[str] = list(self._group_fields) + self._aggregates
        if not select_parts:
            select_parts = ["*"]

        sql = f"SELECT {', '.join(select_parts)} FROM {table}"

        if self._where_clauses:
            sql += " WHERE " + " AND ".join(self._where_clauses)

        if self._group_fields:
            sql += " GROUP BY " + ", ".join(self._group_fields)

        if self._having_clauses:
            sql += " HAVING " + " AND ".join(self._having_clauses)

        if self._order_clauses:
            sql += " ORDER BY " + ", ".join(self._order_clauses)

        if self._limit_val is not None:
            sql += f" LIMIT {self._limit_val}"

        return sql + ";"

    # -- internal -----------------------------------------------------------

    def _run_sql(self, sql: str) -> Any:
        if self._session is not None:
            return self._session.query(sql)
        orm = self._model_cls._get_orm()
        if orm is None:
            raise QueryError("Model is not bound to a VedaORM instance")
        return orm.query(sql)

File: node/vedadb-orm/src/aggregation.ts
/**
 * VedaDB ORM AggregationBuilder
 */

import { escapeIdentifier, escapeValue } from './utils';
import { SchemaDefinition } from './schema';
import { DriverResult } from './types';

// ---------------------------------------------------------------------------
// AggregationBuilder
// ---------------------------------------------------------------------------

interface AggregateExpression {
  sql: string;
  alias: string;
}

interface HavingCondition {
  sql: string;
}

export class AggregationBuilder<T = any> {
  private table: string;
  private schema: SchemaDefinition<T>;
  private executor: (sql: string) => Promise<DriverResult>;

  private groupByFields: string[] = [];
  private havingConditions: HavingCondition[] = [];
  private aggregates: AggregateExpression[] = [];
  private whereSQL: string | null = null;

  constructor(
    schema: SchemaDefinition<T>,
    executor: (sql: string) => Promise<DriverResult>,
    whereSQL?: string
  ) {
    this.table = schema.tableName;
    this.schema = schema;
    this.executor = executor;
    this.whereSQL = whereSQL || null;
  }

  /**
   * Group results by one or more fields.
   */
  groupBy(...fields: (keyof T & string)[]): AggregationBuilder<T> {
    this.groupByFields.push(...fields);
    return this;
  }

  /**
   * Add a HAVING condition.
   */
  having(field: string, op: string, value: any): AggregationBuilder<T> {
    this.havingConditions.push({
      sql: `${field} ${op} ${escapeValue(value)}`,
    });
    return this;
  }

  /**
   * COUNT aggregate.
   */
  count(field: string = '*', alias: string = 'count'): AggregationBuilder<T> {
    const col = field === '*' ? '*' : escapeIdentifier(field);
    this.aggregates.push({ sql: `COUNT(${col})`, alias });
    return this;
  }

  /**
   * SUM aggregate.
   */
  sum(field: keyof T & string, alias?: string): AggregationBuilder<T> {
    this.aggregates.push({
      sql: `SUM(${escapeIdentifier(field)})`,
      alias: alias || `sum_${field}`,
    });
    return this;
  }

  /**
   * AVG aggregate.
   */
  avg(field: keyof T & string, alias?: string): AggregationBuilder<T> {
    this.aggregates.push({
      sql: `AVG(${escapeIdentifier(field)})`,
      alias: alias || `avg_${field}`,
    });
    return this;
  }

  /**
   * MIN aggregate.
   */
  min(field: keyof T & string, alias?: string): AggregationBuilder<T> {
    this.aggregates.push({
      sql: `MIN(${escapeIdentifier(field)})`,
      alias: alias || `min_${field}`,
    });
    return this;
  }

  /**
   * MAX aggregate.
   */
  max(field: keyof T & string, alias?: string): AggregationBuilder<T> {
    this.aggregates.push({
      sql: `MAX(${escapeIdentifier(field)})`,
      alias: alias || `max_${field}`,
    });
    return this;
  }

  /**
   * Generate the SQL.
   */
  toSQL(): string {
    const selectParts: string[] = [];

    // Group by fields go in SELECT
    for (const f of this.groupByFields) {
      selectParts.push(escapeIdentifier(f));
    }

    // Aggregate expressions
    for (const agg of this.aggregates) {
      selectParts.push(`${agg.sql} AS ${escapeIdentifier(agg.alias)}`);
    }

    if (selectParts.length === 0) {
      selectParts.push('COUNT(*) AS "count"');
    }

    let sql = `SELECT ${selectParts.join(', ')} FROM ${escapeIdentifier(this.table)}`;

    if (this.whereSQL) {
      sql += ` WHERE ${this.whereSQL}`;
    }

    if (this.groupByFields.length > 0) {
      sql += ' GROUP BY ' + this.groupByFields.map(escapeIdentifier).join(', ');
    }

    if (this.havingConditions.length > 0) {
      sql += ' HAVING ' + this.havingConditions.map((h) => h.sql).join(' AND ');
    }

    sql += ';';
    return sql;
  }

  /**
   * Execute the aggregation query.
   */
  async exec(): Promise<Record<string, any>[]> {
    const sql = this.toSQL();
    const result = await this.executor(sql);
    return result.toObjects();
  }
}
